File: test/src/common/commonutil/java/TCPAndroidClient/src/main/java/org/iotivity/tcpandroidclient/TCPClient.java
/******************************************************************
 * Copyright 2017 Samsung Electronics All Rights Reserved.
 *
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************/

package org.iotivity.tcpandroidclient;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.net.InetAddress;
import java.net.Socket;
import android.util.Log;

import org.iotivity.configuration.IConfiguration;

public class TCPClient implements IConfiguration {

    private String            serverMessage;
    private OnMessageReceived mMessageListener = null;
    private boolean           mRun             = false;

    private PrintWriter       mOut              = null;
    private BufferedReader    mIn               = null;

    /**
     * Constructor of the class. OnMessagedReceived listens for the messages
     * received from server
     */
    public TCPClient(final OnMessageReceived listener) {
        mMessageListener = listener;
    }

    /**
     * Sends the message entered by client to the server
     *
     * @param message
     *            text entered by client
     */
    public void sendMessage(String message) {
        if (mOut != null && !mOut.checkError()) {
            System.out.println("message: " + message);
            mOut.println(message);
            mOut.flush();
        }
    }

    public void stopClient() {
        mRun = false;
    }

    public void run() {

        mRun = true;

        try {
            InetAddress serverAddr = InetAddress.getByName(SERVERIP);
            Socket socket = new Socket(serverAddr, SERVERPORT);
            try {

                mOut = new PrintWriter(new BufferedWriter(
                        new OutputStreamWriter(socket.getOutputStream())),
                        true);
                mIn = new BufferedReader(
                        new InputStreamReader(socket.getInputStream()));

                while (mRun) {
                    serverMessage = mIn.readLine();

                    if (serverMessage != null && mMessageListener != null) {
                        mMessageListener.messageReceived(serverMessage);
                        Log.e("RESPONSE FROM SERVER",
                                "S: Received Message: '" + serverMessage + "'");
                    }
                    serverMessage = null;
                }
            } catch (Exception e) {
                e.printStackTrace();
            } finally {
                socket.close();
            }

        } catch (Exception e) {

            Log.e("TCP SI Error", "SI: Error", e);

        }

    }

    public interface OnMessageReceived {
        public void messageReceived(String message);
    }
}

File: test/src/testapp/TcpClientApp/tcpclient.cpp
/******************************************************************
*
* Copyright 2017 Samsung Electronics All Rights Reserved.
*
*
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
******************************************************************/

#include "Configuration.h"
#include "CommonUtil.h"
#include "IotivityTest_Logger.h"

const int BUFFER_LENGTH = 1024;
Configuration g_configPropFile(COMMON_CONFIG_FILE);
string g_tcpServerIp = "127.0.0.1";
int g_port = 0;

void killConnection(char *errorMessage)
{
    perror(errorMessage);
    exit(1);
}

int main(int argc, char *argv[])
{
    int socketDescriptor, nunmberOfBytes; /* Socket descriptor */
    struct sockaddr_in tcpServer; /* Echo server address */
    char buffer[BUFFER_LENGTH];
    int g_port = stoi(g_configPropFile.getProperties(TCP_SERVER_PORT_KEY));

#ifdef __LINUX__
    g_tcpServerIp = g_configPropFile.getProperties(TCP_SERVER_ETHERNET_IP_KEY);
#elif __TIZEN__
    g_tcpServerIp = g_configPropFile.getProperties(TCP_SERVER_WIFI_IP_KEY);
#endif

    /* Create a reliable, stream socket using TCP */
    if ((socketDescriptor = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP)) < 0) {
        killConnection("socket() failed");
    }

    /* Construct the server address structure */
    memset(&tcpServer, 0, sizeof(tcpServer));
    tcpServer.sin_family = AF_INET;
    tcpServer.sin_addr.s_addr = inet_addr(g_tcpServerIp.c_str());
    tcpServer.sin_port = htons(g_port);

    /* Establish the connection to the tcp server */
    if (connect(socketDescriptor, (struct sockaddr *) &tcpServer, sizeof(tcpServer)) < 0)
    {
        killConnection("connect() failed");
    }
    else
    {
        IOTIVITYTEST_LOG(INFO, "[TCP Client] Connection Successful ");
    }

    while(true){
        IOTIVITYTEST_LOG(INFO, "[TCP Client] Enter Message : ");
        bzero(buffer, BUFFER_LENGTH);
        fgets(buffer, BUFFER_LENGTH-1, stdin);

        /* Send message to the server */
        nunmberOfBytes = write(socketDescriptor, buffer, strlen(buffer));

        if (nunmberOfBytes < 0)
        {
            IOTIVITYTEST_LOG(ERROR, "[TCP Client] Error writing to socket");
            exit(1);
        }
    }

    IOTIVITYTEST_LOG(INFO, "[TCP Client] Exiting TCP Client : ");
    return 0;
}
